### DISCUSSION TITLE
Implementing compose spec directly

### DISCUSSION THREAD

**Discussion by Aposhian** [containers/podman#13425] [Ideas]
As mentioned in https://github.com/containers/podman/discussions/10976, it is an ongoing effort to maintain compatibility with an evolving Docker Compose. However, now that there is the generic [compose-spec](https://github.com/compose-spec/compose-spec), we could potentially help users to use existing `docker-compose.yaml` files without worrying about providing compatibility with the Docker API.

related: https://github.com/containers/podman/issues/11780

**Comment by Aposhian** [comment]
My background is that I have an application that is defined with `docker-compose.yaml` files. It is desirable to leave it in that format since many developers have Docker, and it is a familiar development tool. However, for certain environments, it would be nice to run the same application, but with Podman. Eventually, I would love to be able to manage those individual services with systemd, but using the forking model of Podman vs the client-server architecture of Docker is a first step.

Right now, users have two main options for this in the Podman ecosystem:
1. `podman-compose`, which parses yaml files, and then runs `podman` CLI commands.
    - Pros:
      - No podman socket is required
    - Cons:
      - It doesn't use the `compose-go` library to parse compose YAML files, so that makes it more difficult to maintain as the spec evolves
      - Being a python script doesn't make it as pluggable into other parts of the podman libraries, such as systemd unit generation (which is what I am most interested in the long run)
      - The code is harder to read since it is executing a CLI from Python, versus calling library functions
3. Using `docker-compose`, but communicating with a `podman` socket over the Docker API.
    - Pros:
        - Compose YAML files should be parsed correctly
    - Cons:
        - Requires a podman socket to be running.
        - Compatibility is dependent on podman keeping up with the Docker API

What I am proposing is that we add a library to Podman that does the following:
- Uses `compose-go` to load yaml files, and convert them into a collection of `libpod.Container` and/or `libpod.Pod`
- Add a `compose` verb to the Podman CLI that loads YAML files and exposes a various functionalities to create, start, run, and stop containers, etc. Keep in mind that while it could be a good idea for `podman compose` to have similar options to `docker-compose` and `docker compose`, it needn't necessarily have the same options. The Compose spec defines the format of the files, not the format of the CLI: it doesn't require the implementing program to have "up" and "down" verbs, for example.

**Comment by baude** [comment]
Thanks @Aposhian ... before you start work on this, I would suggest we get more of the team to comment; moreover, I think a community cabal would be appropriate for just this thing. @TomSweeneyRedHat could you add this to the next meeting and share here when that is as well as get it on the agenda?

**Comment by afbjorklund** [comment]
@Aposhian including support for compose V2 (Go version) was the path taken by `nerdctl compose`.
It didn't _have_ a docker socket, so going with compose V1 (Py version) was never an option, but anyway...

However, it is a long journey to make it compatible:
https://github.com/containerd/nerdctl#compose

Writing directly to the spec (https://compose-spec.io/) seems like the best, long term. Chosen for containerd.
But I think that if you ask the users they would ask What Would Docker Do. So you end up with daemons etc.

So it seems like options 2. followed by 1. are most "likely".
Maybe it would be possible to share code, for option 3. ?

**Reply by Aposhian** [reply]
I don't have code yet that accomplishes option 3, but that is what I am proposing to develop. Yes, I agree that full compose-spec compatibility is a lot, but I think that at least in my experience, a minority of features make up a majority of use-cases. The compose spec also notes:
> We acknowledge that no Compose implementation is expected to support all attributes, and that support for some properties is Platform dependent and can only be confirmed at runtime.

**Reply by afbjorklund** [reply]
Looks like your best option is to rewrite `podman-compose` in Go ?

**Comment by Luap99** [comment]
There is already confusion if users should use podman-compose or docker-compose. Adding a new implementation directly into podman will make it even worse. All of the tools will end up having differences so migrating between them may not be possible.

I also do not think we have the manpower to maintain this. Even if someone would do the initial work we would still need to maintain this in the long run (fixing bugs and keep it up with the compose spec). There is a significant amount of work here just to reimplement already working functionality via podman/docker-compose, so why not just improve these tools?

**Reply by Aposhian** [reply]
From my perspective, both `podman-compose`, and Docker API compatibility seem to be harder to maintain than implementing the compose spec directly. `podman-compose` is implemented ad-hoc, without a backing library like `compose-go` or way to keep its schema in lockstep. The difficulty of trying to maintain compatibility with the Docker API is at the whims of Docker and the features they add, which may not be aligned with the goals or capabilities of Podman. For example, I don't think it is reasonable for Podman to implement Docker Swarm features, correct? Whereas if the compose spec is implemented directly, Podman can more directly choose which features of the spec it will support, and which it will not. I don't think I would ever anticipate full compatibility with docker compose for advanced users, but I think that is expected, given that Podman and Docker have different feature sets and different paradigms.

**Reply by afbjorklund** [reply]
After the addition of Podman Daemon (instead of varlink) and Podman Desktop (instead of libmachine), I would actually not be **too** surprised to see Podman Swarm added - but I think that speculation is _well_ outside this Podman Compose discussion. 🙂

Note that `docker` does **not** support compose... That is handled by an external  `docker-compose` CLI plugin. Maybe that could be an approach also for your `podman-compose`, instead of adding it to the main project ? Have it as a separate binary.

But it seems that adding the compose-spec support is "competing" with the current `podman play kube` approach ? (Kompose?)

Or that "all" that is needed here is to rewrite `podman-compose` in Go instead of Python, so that it can be linked with the rest of it.

**Reply by Luap99** [reply]
We already decided to support the docker api. There are a huge amount of applications using the docker api and we want to support this with podman. docker-compose is just one of them, so it does not really add more maintenance burden on us, in fact, if it will fail it helps us to find bugs in the compat api.

Podman-compose is a community project, there is not much extra maintenance for the podman developers, except for some bugs or feature request that would be useful for podman-compose.

Yes we do not support swarm but I do not see how implementing the spec directly will help here.

IMO users do not care about specs. Instead they are using an application (docker compose) and would expect a `podman compose` to work exactly the same.

**Reply by afbjorklund** [reply]
That is fine for V1 support, but doesn't help for V2 support ?

Like with "buildx" plugin, there _will_ be more requests coming from the "compose" plugin... Which won't be covered by the older specs.

**Reply by Aposhian** [reply]
>Note that docker does not support compose... That is handled by an external docker-compose CLI plugin. Maybe that could be an approach also for your podman-compose, instead of adding it to the main project ? Have it as a separate binary.

Being rather new to Podman, I didn't know that Podman supported CLI plugins. I would be fine to work on this as a separate CLI plugin for `podman`, essentially as a Go rewrite of `podman-compose`.

>But it seems that adding the compose-spec support is "competing" with the current podman play kube approach ? (Kompose?)

I haven't used `podman play kube` before, but looking at it, it seems sort of possible to achieve what I want by doing using Kompose to convert a `docker-compose.yaml` into Kubernetes manifests, and then running them with `podman play kube`. However, it does seem like a little bit of an anti-pattern to convert to Kubernetes manifests when Kubernetes isn't actually used at all (if I understand `play kube` correctly).

>IMO users do not care about specs. Instead they are using an application (docker compose) and would expect a podman compose to work exactly the same.

As a user of `docker-compose`, what care about most is compatibility with the YAML files, so that they can be run by different programs in different ways.

Also, to be fair, the main reason why I would like to rewrite `podman-compose` in Go is to make its logic more reusable. My eventual goal is to run docker-compose files with Podman, where each container is managed as a systemd service. I could just duplicate that logic, but I would like it be something that `podman-compose` can share. I can take a crack at it, and see what the other `podman-compose` developers think of my code. :smile:

**Reply by afbjorklund** [reply]
> I didn't know that Podman supported CLI plugins.

I don't think that it has plugins - yet...

But it could be an alternative, if not accepting more code into the project but yet wanting to offer a `podman compose` user experience ?

**Reply by mheon** [reply]
We'd like to implement them, but Docker upstream deprecated them and pulled the specs. They still *use* them, though there seems to be discussions about changing them.

We'll probably still end up implementing them, assuming they aren't completely removed before we get around to it...

**Reply by mheon** [reply]
Regarding implementing Compose directly in Podman, I share the concerns of the other maintainers, but it really depends on how much code it ends up being. If it ends up being quite easy to implement 80% of the functionality (that will probably end up being ~95% of usecases) then I wouldn't mind merging that. Meanwhile, if this ends up being thousands of lines to implement more like 50% of the functionality, I get a lot more hesitant.

**Reply by Aposhian** [reply]
Well, I could try an initial Go implementation, maybe trying to enlist some help from contributors at `podman-compose`, and then it would be easier to evaluate whether or not it should belong in the main Podman codebase?

**Reply by muayyad-alsadi** [reply]
@Aposhian My personal preference is to have `podman-compose` registered as a CLI plugin outside `podman` code base. That CLI plugin can be written in any language (go or python). Currently I'm going for python because of how fast I can iterate and because I want to keep it process model (not remote/socket API). I'll keep an eye on any GO implementation and if you need any help there just ping me and I'll help you port it.
